DAO.sol

```diff
@@ -511,9 +535,56 @@ contract DAO is DAOInterface, Token, TokenCreation {
             blocked[msg.sender] = _proposalID;
         }
 
+        votingRegister[msg.sender].push(_proposalID);
         Voted(_proposalID, _supportsProposal, msg.sender);
     }
 
+    function unVote(uint _proposalID){
+        Proposal p = proposals[_proposalID];
+
+        if (now >= p.votingDeadline) {
+            throw;
+        }
+
+        if ((p.votedYes[msg.sender]
+            || p.votedNo[msg.sender])
+            && now > p.votingDeadline - preSupportTime) {
+            throw;
+        }
+
+        if (p.votedYes[msg.sender]) {
+            p.yea -= balances[msg.sender];
+            p.votedYes[msg.sender] = false;
+        }
+
+        if (p.votedNo[msg.sender]) {
+            p.nay -= balances[msg.sender];
+            p.votedNo[msg.sender] = false;
+        }
+    }
+
+    function unVoteAll() {
+        for (uint i = 0; i < votingRegister[msg.sender].length; i++) {
+            Proposal p = proposals[i];
+            if (now < p.votingDeadline)
+                unVote(i);
+        }
+
+        votingRegister[msg.sender].length = 0;
+        blocked[msg.sender] = 0;
+    }
+
+    function verifyPreSupport(uint _proposalID) {
+        Proposal p = proposals[_proposalID];
+        if (now < p.votingDeadline - preSupportTime) {
+            if (p.yea > p.nay) {
```

Comment:
@cougarously You are correct.
@CJentzsch The unvote function needs the proposals to keep how much each user voted.
So we need:

```js
        // Simple mapping to check if a shareholder has voted for it
        mapping (address => bool) votedYes;
        // Simple mapping to check if a shareholder has voted against it
        mapping (address => bool) votedNo;
```

to change to:

```js
        // Simple mapping to check if a shareholder has voted for it
        mapping (address => int) votedYesAmount;
        // Simple mapping to check if a shareholder has voted against it
        mapping (address => int) votedNoAmount;
```

in order to be able to keep the original voting state in case someone else transfers tokens to or from the voter's address.